src/testdrive/src/action/consistency.rs

```diff
@@ -0,0 +1,239 @@
+use serde::{Deserialize, Serialize};
+
+use crate::action::{ControlFlow, State};
+use crate::parser::BuiltinCommand;
+
+/// Level of consistency checks we should enable on a testdrive run.
+#[derive(clap::ValueEnum, Default, Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Level {
+    /// Run the consistency checks after the completion of a test file.
+    #[default]
+    File,
+    /// Run the consistency checks after each statement, good for debugging.
+    Statement,
+    /// Disable consistency checks entirely.
+    Disable,
+}
+
+impl FromStr for Level {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "file" => Ok(Level::File),
+            "statement" => Ok(Level::Statement),
+            "disable" => Ok(Level::Disable),
+            s => Err(format!("Unknown consistency check level: {s}")),
+        }
+    }
+}
+
+/// Runs consistency checks against multiple parts of Materialize to make sure we haven't violated
+/// our invariants or leaked resources.
+pub async fn run_consistency_checks(state: &State) -> Result<ControlFlow, anyhow::Error> {
+    let coordinator = check_coordinator(state).await.context("coordinator");
+    let catalog_state = check_catalog_state(state).await.context("catalog state");
+    // TODO(parkmycar): Fix subsources so they don't leak their shards and then add a leaked shards
+    // consistency check.
+
+    // Make sure to report all inconsistencies, not just the first.
+    let mut msg = String::new();
+    if let Err(e) = coordinator {
+        writeln!(&mut msg, "coordinator inconsistency: {e:?}")?;
+    }
+    if let Err(e) = catalog_state {
+        writeln!(&mut msg, "catalog inconsistency: {e:?}")?;
+    }
+
+    if msg.is_empty() {
+        Ok(ControlFlow::Continue)
+    } else {
+        Err(anyhow!("{msg}"))
+    }
+}
+
+/// Checks if a shard in Persist has been tombstoned.
+///
+/// TODO(parkmycar): Run this as part of the consistency checks, instead of as a specific command.
+pub async fn run_check_shard_tombstoned(
+    mut cmd: BuiltinCommand,
+    state: &State,
+) -> Result<ControlFlow, anyhow::Error> {
+    let shard_id = cmd.args.string("shard-id")?;
+    check_shard_tombstoned(state, &shard_id).await?;
+    Ok(ControlFlow::Continue)
+}
+
+/// Asks the Coordinator to run it's own internal consistency checks.
+async fn check_coordinator(state: &State) -> Result<(), anyhow::Error> {
+    let response = Retry::default()
+        .max_duration(Duration::from_secs(2))
+        .retry_async(|_| async {
+            reqwest::get(&format!(
+                "http://{}/api/coordinator/check",
+                state.materialize_internal_http_addr,
+            ))
+            .await
+        })
+        .await
+        .context("querying coordinator")?;
+    if response.status() == StatusCode::NOT_FOUND {
+        tracing::info!("Coordinator consistency check not available");
+        return Ok(());
+    }
```

Comment:
When do we expect this to happen?

Reply:
We don't ever expect this to happen, let me upgrade them to errors and see what happens

Reply:
Yeah, it's probably better to upgrade these to errors. That way if someone misconfigures a test and these aren't run then they get an error.